crates/pink-libs/subrpc/src/lib.rs

```diff
@@ -337,6 +341,39 @@ pub fn send_transaction(rpc_node: &str, signed_tx: &[u8]) -> core::result::Resul
     hex::decode(&resp.result[2..]).or(Err(Error::InvalidBody))
 }
 
+
+/// Query a wasm smart contract
+pub fn query_contract<E: scale::Decode, B: scale::Decode>(rpc_node: &str, query: &[u8], at: Option<H256>) -> Result<ContractQueryResult<E, B>> {
+
+    let query_hex = format!("0x{}", hex::encode(query));
+    let maybe_hex_at = at.map_or("null".to_string(), |h| format!("\"0x{h:x}\""));
+
+    let query = format!(
+        r#"{{"id":1,"jsonrpc":"2.0","method":"state_call","params":["ContractsApi_call", "{query_hex}", {maybe_hex_at}]}}"#
+    );
+
+    let resp_body = call_rpc(rpc_node, query.into_bytes())?;
+    let resp: QueryContractResponse = json::from_slice(&resp_body).or(Err(Error::InvalidBody))?;
+
+    let result = match resp.result {
+        Some(h) => hex::decode(&h[2..]).or(Err(Error::InvalidBody))?,
+        None => return Err(Error::InvalidBody),
+    };
+
+    let contract_query_result = <ContractQueryResult<E, B>>::decode(&mut result.as_slice())
+        .map_err(|_| Error::DecodeFailed)?;
+
+    if contract_query_result.result.flags != 0 {
+        // it means the contract returns an error
+        return Err(Error::ContractError);
```

Comment:
As a library, it is preferable not to perform error conversions here, as this could result in the loss of detailed error information.

Reply:
I agree, I updated the code.